env/io_posix.h

```diff
@@ -31,39 +36,81 @@
 #endif
 
 namespace ROCKSDB_NAMESPACE {
-static std::string IOErrorMsg(const std::string& context,
-                              const std::string& file_name) {
-  if (file_name.empty()) {
-    return context;
+std::string IOErrorMsg(const std::string& context,
+                       const std::string& file_name);
+// file_name can be left empty if it is not unkown.
+IOStatus IOError(const std::string& context, const std::string& file_name,
+                 int err_number);
+
+class PosixHelper {
+ public:
+  static size_t GetUniqueIdFromFile(int fd, char* id, size_t max_size);
+  static size_t GetLogicalBufferSize(int fd);
+};
+
+#ifdef OS_LINUX
+// Files under a specific directory have the same logical buffer size.
+// This class caches the logical buffer size for the specified directories to
+// save the CPU cost of computing the size.
+// Safe for concurrent access from multiple threads without any external
+// synchronization.
+class LogicalBufferSizeCache {
+ public:
+  LogicalBufferSizeCache(std::function<size_t(int)> get_logical_buffer_size =
+                             PosixHelper::GetLogicalBufferSize)
+      : get_logical_buffer_size_(get_logical_buffer_size) {}
+
+  // Logical buffer size of files under each specified directory will be cached.
+  // If the logical buffer size for a directory has been cached, the original
+  // cached size will keep being used.
+  void AddCacheDirectories(const std::unordered_set<std::string>& directories) {
+    WriteLock lock(&cache_mutex_);
+    for (const auto& d : directories) {
+      auto& s = cache_[d];
+      (void) s;
```

Comment:
Why don't we call get_logical_buffer_size_() here so that we don't need to call it in size()?

Reply:
I was concerning of potentially large number of dbpaths, so get_logical_buffer_size_ was called on demand. But since number of dbpaths in practice is not huge, I updated the PR to call  get_logical_buffer_size_ when adding the dbpaths.